vscode-dotty/src/tracer.ts

```diff
@@ -0,0 +1,370 @@
+                log += value
+            },
+
+            appendLine: (value: string) => {
+                localOutputChannel.appendLine(value)
+                if (this.tracingConsent.get() === 'no') {
+                    log = ''
+                    return
+                }
+
+                log += value
+                log += '\n'
+                if (this.tracingConsent.get() === 'yes') withSocket((socket) => {
+                    if (socket.readyState === WebSocket.OPEN) {
+                        const send = (msg: string) => socket.send(msg, (err) => {
+                            if (err) {
+                                this.logError('socket send error', err)
+                            }
+                        })
+
+                        let start = 0
+                        while (start < log.length) {
+                            send(log.substring(start, start + this.maximumMessageSize))
+                            start += this.maximumMessageSize
+                        }
+                        log = ''
+                    }
+                })
+            },
+
+            clear() { },
+            show() { },
+            hide() { },
+            dispose() {
+                if (socket) socket.close()
+                localOutputChannel.dispose()
+            }
+        }
+    }
+
+    private silenceErrors: boolean = false
+    private logErrorWithoutNotifying(message: string, ...rest: any[]) {
+        const msg = `[Dotty LSP Tracer] ${message}`
+        // unwrap SafeJsonifier, for some reason Electron logs the result
+        // of .toJSON, unlike browsers
+        console.error(msg, ...rest.map((a) => a instanceof SafeJsonifier ? a.value : a))
+        function cautiousStringify(a: any): string  {
+            try {
+                return JSON.stringify(a, undefined, 4)
+            } catch (err) {
+                console.error('cannot stringify', err, a)
+                return a.toString()
+            }
+        }
+        this.ctx.extensionOut.appendLine([msg].concat(rest.map(cautiousStringify)).join(' '))
+    }
+    private logError(message: string, ...rest: any[]) {
+        this.logErrorWithoutNotifying(message, ...rest)
+        if (!this.silenceErrors) {
+            vscode.window.showErrorMessage(
+                'An error occured which prevents sending usage data to EPFL. ' +
+                'Please copy the text from "Dotty Language Client" output (View > Output) and send it to your TA.',
+                'Silence further errors'
+            ).then((result) => {
+                if (result !== undefined) {
+                    this.silenceErrors = true
+                }
+            })
+        }
+    }
+}
+
+function safeError(e: Error): SafeJsonifier<Error> {
+    return new SafeJsonifier(e, (e) => e.toString())
+}
+
+class SafeJsonifier<T> {
+    value: T
+    valueToObject: (t: T) => {}
+
+    constructor(value: T, valueToObject: (t: T) => {}) {
```

Comment:
You can directly create an immutable class field and assign by writing `constructor(readonly value: T, ...)`